kotlinx-coroutines-core/common/src/flow/internal/Combine.kt

```diff
@@ -9,129 +9,141 @@ import kotlinx.coroutines.*
+    var currentEpoch: Byte = 0
+    while (true) {
+        ++currentEpoch
+        // Start batch
+        // The very first receive in epoch should be suspending
+        var element = resultChannel.receiveOrNull() ?: break // Channel is closed, nothing to do here
+        while (true) {
+            val index = element.index
+            // Update values
+            val previous = latestValues[index]
+            latestValues[index] = element.value
+            if (previous === UNINITIALIZED) --remainingAbsentValues
+            // Check epoch
+            // Received the second value from the same flow in the same epoch -- bail out
+            if (lastReceivedEpoch[index] == currentEpoch) break
+            lastReceivedEpoch[index] = currentEpoch
+            element = resultChannel.poll() ?: break
+        }
 
-// Channel has any type due to onReceiveOrNull. This will be fixed after receiveOrClosed
-private fun CoroutineScope.asFairChannel(flow: Flow<*>): ReceiveChannel<Any> = produce {
-    val channel = channel as ChannelCoroutine<Any>
-    flow.collect { value ->
-        return@collect channel.sendFair(value ?: NULL)
+        // Process batch result if there is enough data
+        if (remainingAbsentValues == 0) {
+            val results = arrayFactory()
+            (latestValues as Array<T?>).copyInto(results)
+            transform(results as Array<T>)
+        }
     }
 }
 
-internal fun <T1, T2, R> zipImpl(flow: Flow<T1>, flow2: Flow<T2>, transform: suspend (T1, T2) -> R): Flow<R> = unsafeFlow {
-    coroutineScope {
-        val first = asChannel(flow)
-        val second = asChannel(flow2)
-        /*
-         * This approach only works with rendezvous channel and is required to enforce correctness
-         * in the following scenario:
-         * ```
-         * val f1 = flow { emit(1); delay(Long.MAX_VALUE) }
-         * val f2 = flowOf(1)
-         * f1.zip(f2) { ... }
-         * ```
-         *
-         * Invariant: this clause is invoked only when all elements from the channel were processed (=> rendezvous restriction).
-         */
-        (second as SendChannel<*>).invokeOnClose {
-            if (!first.isClosedForReceive) first.cancel(AbortFlowException(this@unsafeFlow))
-        }
+internal fun <T1, T2, R> zipImpl(flow: Flow<T1>, flow2: Flow<T2>, transform: suspend (T1, T2) -> R): Flow<R> =
+    unsafeFlow {
+        coroutineScope {
+            val second = asChannel(flow2)
+            /*
+             * This approach only works with rendezvous channel and is required to enforce correctness
+             * in the following scenario:
+             * ```
+             * val f1 = flow { emit(1); delay(Long.MAX_VALUE) }
+             * val f2 = flowOf(1)
+             * f1.zip(f2) { ... }
+             * ```
+             *
+             * Invariant: this clause is invoked only when all elements from the channel were processed (=> rendezvous restriction).
+             */
+            val collectJob = Job()
```

Comment:
The way this is written now would not properly display in coroutines debugger when the first flow is being collector. It should be, at least, changed to `val collectJob = Job(coroutineContext[Job])`, so it is a child of the current job. But, in fact, it should use an implementation of the `Job` that inherits `ScopeCoroutine`.

Reply:
Good observation, tho it works perfectly fine in the debugger right now because `withContextUndispatched` creates no coroutine and frame walker successfully finds the last suspension point.

Adding `Job(coroutineContext[Job])` will make things worse as standalone and irrelevant `JobImpl` will be printed by `DebugProbes.printJob`

Reply:
Currently it does not really work if you put breakpoint inside `collect`. Checkout this test in debugger:

```
package kotlinx.coroutines

import kotlinx.coroutines.flow.*
import kotlin.test.*

class ScratchTest : TestBase() {
    @Test
    fun test() = runTest {
        val f1 = flowOf("a")
        val f2 = flowOf(1)
        f1.zip(f2) { i, j -> i + j }.collect {
            foo()
        }
    }
}

suspend fun foo() {
    yield() // suspend
    println("!!!") // breakpoint
}
```

You see the 1st coroutine in the debugger, but it does not show the correct suspention point inside `foo`, because the `Job` that it is suspended inside is detached.

Reply:
Aha! I was thinking about the other job.

>But, in fact, it should use an implementation of the Job that inherits ScopeCoroutine

Won't really help because stackwalking doesn't leverage jobs in any way (otherwise `withContext(Job())` wouldn't work)
Neither won't help `Job(coroutineContext[Job])`.

Another observation is that we have broken stackwalking in multiple places: in `zip` and, for the same reason, in `flowOn`. Also, while writing tests for "the same as zip, but flowOn", I've uncovered a bug in `SafeCollector` with the same cause.
Luckily they all have straightforward fix